File: apps/ios/GalaxySSI/AgentTranscriptScrollPolicy.swift
import Foundation

enum AgentTranscriptScrollPolicy {
  static func nextAutoFollow(
    current: Bool,
    userScrollActive: Bool,
    itemCount: Int,
    lastVisiblePosition: Int,
    remainingPx: Int,
    thresholdPx: Int
  ) -> Bool {
    if !userScrollActive {
      return current
    }
    return itemCount == 0 ||
      (lastVisiblePosition == itemCount - 1 && remainingPx <= thresholdPx)
  }

  static func shouldLoadOlderFromScroll(
    dy: Int,
    firstVisiblePosition: Int,
    hydrationPending: Bool
  ) -> Bool {
    dy < 0 &&
      !hydrationPending &&
      firstVisiblePosition <= 1
  }

  static func shouldLoadOlderFromPull(
    downY: Double,
    currentY: Double,
    canScrollUp: Bool,
    hydrationPending: Bool,
    thresholdPx: Int
  ) -> Bool {
    !hydrationPending &&
      !canScrollUp &&
      currentY - downY >= Double(thresholdPx)
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentTranscriptScrollPolicy.kt
package com.galaxyssi.chat

internal object AgentTranscriptScrollPolicy {
    fun nextAutoFollow(
        current: Boolean,
        userScrollActive: Boolean,
        itemCount: Int,
        lastVisiblePosition: Int,
        remainingPx: Int,
        thresholdPx: Int
    ): Boolean {
        if (!userScrollActive) return current
        return itemCount == 0 ||
            (lastVisiblePosition == itemCount - 1 && remainingPx <= thresholdPx)
    }

    fun shouldLoadOlderFromScroll(
        dy: Int,
        firstVisiblePosition: Int,
        hydrationPending: Boolean
    ): Boolean =
        dy < 0 &&
            !hydrationPending &&
            firstVisiblePosition <= 1

    fun shouldLoadOlderFromPull(
        downY: Float,
        currentY: Float,
        canScrollUp: Boolean,
        hydrationPending: Boolean,
        thresholdPx: Int
    ): Boolean =
        !hydrationPending &&
            !canScrollUp &&
            currentY - downY >= thresholdPx
}
